### ISSUE TITLE
Status on Cohort Definition JSON Schema?

### ISSUE THREAD

**Issue by TheCedarPrince** [OHDSIBr/ATLAS-JSON-Schema#1]
Hi @pmadril ,

I was wondering, what is the status of the JSON Schema for Cohort Definitions within OHDSI? Is this draft you show here accurate or is there more work being done elsewhere? Wondering if the JSON Schema has been standardized yet.

Thanks!

jz

**Comment by pmadril** [comment]
Hi, Jacob
Thank you for your interest in the JSON Schema for Cohort Definitions
<https://github.com/OHDSIBr/ATLAS-JSON-Schema>.

This is an initiative developed by myself and my wife Tereza, with the hope
to be useful for anyone that wants to create tools, other than Atlas, for
Cohort definitions.
It was complete at the time it was developed and I didn't see changes in
the Atlas cohort definition interface, but maybe I'm wrong.
It is described using JSON Schema draft-04, a lower common denominator in
JSON Schema Standards, because it also has the most tools supporting it
(validators, webui generators, etc).
I also made contributions to https://github.com/json-editor/json-editor,
which I use to test the Schema.
Actually, you can copy an Atlas cohort definition and paste in the JSON
Editor, to get a form where you can change values, and then go back to
Atlas.
Anyway, that's not my intended use, Atlas already does a good job as UI,
but I needed a way to test the schema.
My intended use is to be able to have validators which can check a cohort
definition received in an API against the Schema before processing it.
This will enable tools that process cohort definitions.
I'm not aware of initiatives in OHDSI to standardize a cohort definition.
If there is interest in doing it, I hope this schema helps to start the
process.

Thanks a lot for your interest!
Pablo Madril

**Comment by TheCedarPrince** [comment]
Hey @pmadril , 

> This is an initiative developed by myself and my wife Tereza, with the hope to be useful for anyone that wants to create tools, other than Atlas, for Cohort definitions. It was complete at the time it was developed and I didn't see changes in the Atlas cohort definition interface, but maybe I'm wrong.

This is a fantastic resource! I fully agree and was looking around for a specification just for that person -- to develop other tools beyond just OHDSI. Was hoping to create a parser that could understand the Cohort Definition JSON Schema for another project I am developing. I'll definitely use your reference here as a guide in my efforts.

> I'm not aware of initiatives in OHDSI to standardize a cohort definition. If there is interest in doing it, I hope this schema helps to start the process.

Ah that is unfortunate! I think this is a very good idea and hope it could be adopted within the OHDSI community. 

Either way, thanks for sharing on this work and idea! Feel free to close this issue as I was wondering what was the status of this Schema in general (if it had been standardized, where it was being worked on, etc.). 

~ tcp :deciduous_tree:

**Comment by pmadril** [comment]
Very glad to know it's been useful to you. 
Use any JSON Schema parser with this schema and you will get  a full in memory representation to work on. 
JSON Schema already has available a parser for most modern languages, refer to [Implementations](https://json-schema.org/implementations.html) and I'm sure you will find your favourite language there.
The [single file](https://raw.githubusercontent.com/OHDSIBr/ATLAS-JSON-Schema/2.7.4/single/atlas_cohort_definition.json) version is for you, pure JSON Schema without editor enhancements  in a single file (no need of ajax).
I will close this issue after your reply, but feel free to contact me for help. 
I'm very interested in your use of the schema, indeed, you are the first real user of it. 
Let me know about your progress, problems and deficiencies that you find.

Best regards, Pablo

**Comment by pmadril** [comment]
Hi, Jacob

Obviously, as soon as I wrote and post my reply, I discovered that there are no JSON Schema implementations for Julia, which seems to be your favourite language. 
My shame, I apologise  for that. I knew Julia just by reference but never went deeper about it. Now, thanks to you, I knew a bit more about Julia and I got interest to learn more about it.
Julia stands to have native call to C functions and may be the [WJElement](https://github.com/netmail-open/wjelement) is useful to make some tests. It has some support for draft 4. For C++ there are a lot of options.

As far as I know, any draft-7 support library is already backward compatible with draft-4 but I didn't test it.
I'm thinking about to make a draft-7 version of the schema, which seems to be the new lowest common denominator.

Let me know if I can help you, Pablo.
